stages/shell.init: add pattern for env var names

achilleas-k opened the pull request:
Pattern for valid environment variable names as defined in The Open Group Base Specifications Issue 7, 2018 edition IEEE Std 1003.1-2017 (Revision of IEEE Std 1003.1-2008)

https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap08.html

thozza commented on stages/org.osbuild.shell.init:
```diff
@@ -27,7 +27,10 @@ SCHEMA_2 = r"""
-                "key": "string",
+                "key": {
+                  "type": "string",
+                  "pattern": "[a-zA-Z][a-zA-Z0-9_]*"
```
The document that you linked says 
```text
Environment variable names used by the utilities in the Shell and Utilities volume of
POSIX.1-2017 consist solely of uppercase letters, digits, and the <underscore> ( '_' )
from the characters defined in [Portable Character Set]
(https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap06.html#tagtcjh_3)
and do not begin with a digit.
```
So no lowercase letters... I'm not sure what `The name space of environment variable names containing lowercase letters is reserved for applications.` means in the context of this stage 🤔

achilleas-k replied:
I feel it's technically allowed but might not be strictly POSIX for global env.  I'm good with restricting to ALL CAPS if you think it's more correct.

thozza replied:
Unless we already have a requirement to allow lowercase (e.g. from the EAP team), I would stick to all caps. We can always relax it... I'm also OK with keeping it as is if others are also OK with it

achilleas-k replied:
I'll go with all caps.

achilleas-k replied:
All caps now.  Updated tests.  I also added `_` as an allowed character in the first set, which is valid.
I tried to get the test to fail to see if it all works but realised that the pattern doesn't affect the validation.  Is that right?  Did I write the schema wrong?

regexowl replied:
@achilleas-k Are you matching against the whole string or just a partial match? For full match you'll probably need to add anchors `^[A-Z_][A-Z0-9_]*$`

achilleas-k replied:
Same.  Just ran a build now.
```
❯ grep pattern /usr/lib/osbuild/stages/org.osbuild.shell.init
    "patternProperties": {
                  "pattern": "^[A-Z_][A-Z0-9_]*$"

❯ cat ~/tmp/tree/etc/profile.d/test_env
export -VAR="val"
```

achilleas-k replied:
https://github.com/osbuild/osbuild/commit/690a8fc42af46b700b60314b0fc84cd09da96d16

achilleas-k replied:
That might be something specific to object keys though.

achilleas-k replied:
Thanks for the help!

thozza reviewed (commented):
Respect for figuring out the standard that defines the pattern. However I'm not sure it is correct...

achilleas-k commented:
Fixed the patterns and modified the schema to push it one level down for future global options.
Some failure testing to validate the pattern correctness for filenames and variable names:
```
.pipelines[1].stages[1].options.files:
  'test_env/' does not match any of the regexes: '^[a-zA-Z0-9\\.\\-_]{1,250}$'
```
Previously, this was allowed because I made the mistake of defining the range `.-_` instead of escaping the characters.

```
.pipelines[1].stages[1].options.files.test_env.env[0].key:
  '-VAR' does not match '^[A-Z_][A-Z0-9_]*$'

.pipelines[1].stages[1].options.files.test_env.env[0].key:
  'var' does not match '^[A-Z_][A-Z0-9_]*$'
```

New variable name pattern.

thozza reviewed (commented):
LGTM. Just a note that this is a backward incompatible change for any users of the stage, but it is acceptable in this case, because no manifests use it yet.
